### PR TITLE
FEAT: User Auth on Diagnostics Page

### PR DESCRIPTION
**Pull request by robputt**
Add user auth to the local diagnostics pages...

Features
- SQLite DB with ORM Models + Alembic Migrations
- Run DB Migrations upon application start
- Login + Change Password Form Templates
- Refactor templates for better reuse via JINJA2
- Stores password as bcrypt hash
- Logs failed login attempts and locks login form if > 4 attempts in last 10 mins
- Default password entry set to Ethernet mac address of device with colons removed
- Reset password to default using device button press when on password reset page

### CODE REVIEW

**Comment by shawaj** [comment]
@robputt the security thing is below:

Screenshot 2023-01-03 at 00 23 29

and at https://sonarcloud.io/dashboard?id=NebraLtd_hm-diag&pullRequest=441

I can mark it as safe if it is not an issue, but wasn't sure

**Comment by shawaj** [comment]
also @robputt I was thinking - is there any chance we can add some brute-force protection? perhaps by rate limiting password entries?

just thinking if someone accidentally exposes this to the open internet, we should have some protection to stop people from brute forcing

**Comment by robputt** [comment]
> @robputt the security thing is below:
> 
> Screenshot 2023-01-03 at 00 23 29
> 
> and at https://sonarcloud.io/dashboard?id=NebraLtd_hm-diag&pullRequest=441
> 
> I can mark it as safe if it is not an issue, but wasn't sure

@shawaj please mark this as safe in Sonarqube, we already follow this pattern in the other view and it has been accepted in Sonarqube. 👍

**Comment by robputt** [comment]
> also @robputt I was thinking - is there any chance we can add some brute-force protection? perhaps by rate limiting password entries?
> 
> just thinking if someone accidentally exposes this to the open internet, we should have some protection to stop people from brute forcing

@shawaj - yep I can add this

if password_wrong > 5 tries then block login for 15 mins? 

something like this?

**Comment by shawaj** [comment]
@robputt yep exactly 😊

**Comment by robputt** [comment]
> @NebraLtd/tech-support @ChristopherRush once this is merged we will need a support article explaining the functionality and why we have introduced it and possibly a blog/email as well

Don't worry this is still a few days off, I am changing how it works a bit. 🤦

**Comment by shawaj** [comment]
@robputt passing 🎊🎊

**Comment by shawaj** [comment]
FYI @robputt 

https://sonarcloud.io/project/issues?resolved=false&types=CODE_SMELL&pullRequest=441&id=NebraLtd_hm-diag&open=AYWDOkjZ5xpLamRQHPmL

Screenshot_20230105_201818_Chrome

**Comment by robputt** [comment]
> FYI @robputt
> 
> https://sonarcloud.io/project/issues?resolved=false&types=CODE_SMELL&pullRequest=441&id=NebraLtd_hm-diag&open=AYWDOkjZ5xpLamRQHPmL
> 
> Screenshot_20230105_201818_Chrome

These are indeed different tests but Sonarqube is too dumb to understand how the different @patch affects the test outcome. One overrides BALENA_DEVICE_TYPE the other test does not ;-)

**Comment by shawaj** [comment]
I'll see if I can override.

It's also complaining about repeated password strings in tests

**Comment by robputt** [comment]
> I'll see if I can override.
> 
> It's also complaining about repeated password strings in tests

Well that's dumb, cos it should expect password style fixtures in unit tests that are testing password functions :-\

**Comment by robputt** [comment]
@shawaj ready for review I think :-)

**Comment by shawaj** [comment]
@robputt you have access to sonarcloud so you should be able to override by just clicking on the link and signing in with GitHub login

File: .github/workflows/python-tests.yml [hunk 1]
```diff
@@ -27,7 +27,14 @@ jobs:
...
+    - name: Move migrations to expected location
+      run: |
+        mkdir /opt/migrations
+        cp -r migrations /opt/migrations/.
+        cp -r alembic.ini /opt/migrations/.
+        sudo mkdir /var/data
+        sudo chmod 777 /var/data
     - name: Unit tests
       run: |
         pip install pytest
```
Comment (shawaj) [inline .github/workflows/python-tests.yml:40]: Shall we add pytest-cov to test-requirememts and add

--cov=hw_diag --cov=bigquery --cov-fail-under=70

Comment (shawaj) [inline .github/workflows/python-tests.yml:40]: Shall we add pytest-cov to test-requirememts and add

--cov=hw_diag --cov=bigquery --cov-fail-under=70

**Review by shawaj — COMMENTED** [review]
Looks good to me.

Perhaps @MuratUrsavas or @pritamghanghas could take a look too in case I've missed anything?

**Comment by shawaj** [comment]
Oh and maybe a cheeky squash/rebase @robputt 😜

File: .github/workflows/python-tests.yml (hunk 1 above)
Comment (robputt) [reply .github/workflows/python-tests.yml:40]: Generally I would agree, but I think it's out of scope of this PR. Let's raise this in a separate PR if there is a policy change regarding coverage.

**Review by pritamghanghas — COMMENTED** [review]
Looks like for the user, only way to recover forgotten password is reflash the miner or wipe storage from balena. Does the user have capability to wipe miner on his own without raising a request to customer support.

**Comment by robputt** [comment]
> Looks like for the user, only way to recover forgotten password is reflash the miner or wipe storage from balena. Does the user have capability to wipe miner on his own without raising a request to customer support.

Not currently, they would need to ask Nebra support :-(. @shawaj any thoughts how to handle this?

File: .github/workflows/python-tests.yml (hunk 1 above)
Comment (shawaj) [reply .github/workflows/python-tests.yml:40]: I mean, we have that on other repos such as config https://github.com/NebraLtd/hm-config/blob/c580b108ecacedae69b3b0c2954704646d2118a0/.github/workflows/python-tests.yml#L22

But yeah can put in a follow up ticket if that's easier

**Comment by pritamghanghas** [comment]
@shawaj Here is my suggestion, not sure about security implications.
* have a forgot password button on the login screen. When the user presses it, we can take either of the following options:
   * rest to a random password and send that over email to the user.
   * check if the connection is over local lan and allow for reset of password. Someone on the local network can be reasonably trusted.

**Comment by pritamghanghas** [comment]
@kashifpk will know better about identity in dashboard, but i would expect they can be identified by both uuid and mac addresses.

**Comment by shawaj** [comment]
> @shawaj Here is my suggestion, not sure about security implications.
> 
> * have a forgot password button on the login screen. When the user presses it, we can take either of the following options:
>   
>   * reset to a random password and send that over email to the user.
>   * check if the connection is over local lan and allow for reset of password. Someone on the local network can be reasonably trusted.

The issue with email is that we don't have the users email and also we would have to send through a mail server which seems like overkill on the diagnostics.

Agree about the local network thing, but TBH that can probably be spoofed and could also cause issues if someone's network gets hacked. So I think more cumbersome methods that require physical access or a reflash is probably better

**Comment by shawaj** [comment]
> @kashifpk will know better about identity in dashboard, but i would expect they can be identified by both uuid and mac addresses.

The main thing I'm wondering, is if a device that's already listed in the dashboard changes it's Balena uuid but still has the same Mac address and all other info, what will happen? @kashifpk any idea?

**Comment by robputt** [comment]
To be merged and deployed with https://github.com/NebraLtd/hm-config/pull/225

**Comment by shawaj** [comment]
> > @kashifpk will know better about identity in dashboard, but i would expect they can be identified by both uuid and mac addresses.
> 
> The main thing I'm wondering, is if a device that's already listed in the dashboard changes it's Balena uuid but still has the same Mac address and all other info, what will happen? @kashifpk any idea?

This comment still needs resolving @KevinWassermann94 but apart from that I think we are good to go @robputt so approving but should follow this up. Maybe I'll create a new issue on hm-dashboard to follow this question up

**Review by shawaj — APPROVED** [review]

File: .github/workflows/python-tests.yml (hunk 1 above)
Comment (shawaj) [reply .github/workflows/python-tests.yml:40]: Follow up ticket #445

**Comment by shawaj** [comment]
Follow up ticket for the above... https://github.com/NebraLtd/hm-dashboard/issues/1471

**Comment by kashifpk** [comment]
> The main thing I'm wondering, is if a device that's already listed in the dashboard changes it's Balena uuid but still has the same Mac address and all other info, what will happen? @kashifpk any idea?

@shawaj -AFAIK in such case our balena celery task would create a new balena record in balena table. This is being done [here](https://github.com/NebraLtd/hm-dashboard/blob/master/app/balena_helper/tasks.py#L121). This table links with the main devices table using serial number so the FK would still be valid. One problem though would be that within the devices table itself we have a balena_uuid field that never gets updated during this process so that field would contain old/invalid balena uuid.

**Pull request merged**